### DISCUSSION TITLE
Where can I find the legacy EGSnrc version 4 – 2.3.1 from 2010?

### DISCUSSION THREAD

**Discussion by VSpan7** [nrc-cnrc/EGSnrc#746] [Questions] [answered]
I would like to install a previous version of EGSnrc. Actually I want to install v.4 2.3.1 and I have two issues. Firstly, I cannot find this version online and secondly If I had the older version, I don't know which fortran,C and C++ compiler I had to use. Any information would be appreciated. Thank you in advance.

**Comment by ftessier** [comment, marked as answer]
Here is a link to the self-extracting executables for EGSnrc V4 2.3.1 and BEAMnrc V4 2.3.1 (distributed separately back then), for either GNU/Linux or Windows. These files were released in March 2010. There is no information in the release notes about specific compiler versions, so my suggestion is to try it with recent GNU compilers, or else with GNU compilers from the 2010 time frame. Let me know how it turns out!

https://www.dropbox.com/sh/4c4r7ytzw9mb5ka/AACBMm5vJ2icorGuglhutmcTa

**Reply by VSpan7** [reply]
I installed EGSnrc V4 2.3.1 in Linux from the self-extracting executables you shared, using recent GNU compilers and works fine. I tried then to install Epp following the instructions from Epp user manual (https://jonaslippuner.com/files/epp/Epp_User_Manual_1.4.3.pdf), but I got an error when I used the command `make` : 

```
Makefile:43: egspp_.conf : No such file or directory
make: *** No rule to make target 'egspp_.conf'. Stop.
```

Then I created an empty file called egspp_.conf and run `make` again, getting this error this time : 
```
make: *** No rule to make target 'interfaceegs_interface2.h' , needed by 'Epp_.'. Stop.
```

**Reply by ftessier** [reply]
Hmmm. This seems unrelated to Epp. The `egspp_.conf` name indicates that the configuration has no *name*, which comes after the underscore and is set during the configuration. I don't know why this happens at this point, are you given the choice of a configuration name during the installation? If so, try again and enter some name for the configuration.

**Reply by VSpan7** [reply]
I had to activate my current configuration by adding the following lines to my shell resource file : 
`EGS_HOME=/home/panagiotis/egsnrc_mp/`
`EGS_CONFIG=/home/panagiotis/HEN_HOUSE/specs/x86_64-unknown-linux-gnu-gfortran.conf`
`export EGS_HOME EGS_CONFIG`
`. /home/panagiotis/HEN_HOUSE/scripts/egsnrc_bashrc_additions`

Then I tried to `make` Epp but got errors again.

**Comment by VSpan7** [comment]
Thank you very much for your reply. The reason I need this older version is that I'm trying to use a user code that was only tested in this version. The user code is Epp (Easy particle propagation) and when I tried to use it with the newer EGSnrc version, errors occurred, which can be seen in the attached screenshots. Do you happen to know this user code and how can I use it in a recent
EGSnrc version?

I'll try to install this older version from the link you shared, using recent GNU compilers and I'll let you know how it turns out.

Thank you in advance
Panagiotis

**Reply by ftessier** [reply]
I remember Epp (https://jonaslippuner.com/research/epp/), although I have never used it and it was developed by Jonas Lippuner et al. at the University of Manitoba. I remember that the code was very clean and clear, and well built, hence I am confident in could be compiled with a recent version of EGSnrc with minimal adjustments; I will give it a try!

**Reply by ftessier** [reply]
The problem arises because the header `Epp.h` includes the `egs_advanced_application.cpp` source code file, which is not meant to be included in other files. Changes to this `.cpp` file (`is_pegsless` and `checkDeviceFull()`) induce errors for undeclared symbols. I will check if this can be resolved easily, in which case I would advise using the most recent EGSnrc version.

**Reply by ftessier** [reply]
Ok, maybe the Epp code is not so clean after all 😜, because it relies on `#include` of source files instead of deriving the application class properly.  I tried resolving the issue quickly by inserting the `is_pegsless` and `checkDeviceFull()` changes inside the Epp classes, but ended up with

```/bin/ld: Epp_bce4233.o: undefined reference to symbol '_ZNSs4_Rep10_M_destroyERKSaIcE@@GLIBCXX_3.4'```

I remain convinced this can be resolved without too much effort, and ideally the Epp code could be refactored to survive future EGSnrc updates. But this goes beyond what I can do quickly, unless someone else has an idea?

**Reply by VSpan7** [reply]
Hello @ftessier and thank you for the previous replies. Is there a chance that you have an update about Epp code?

Thank you in advance.

**Reply by ftessier** [reply]
@VSpan7 I reached out to Jonas Lippuner (now at Los Alamos National Laboratory), and he agrees that the code he wrote back then was somewhat tangled up with `#include` directives to include`.cpp` files, etc. Jonas would be happy to contribute `Epp` to be distributed with EGSnrc if we can untangled that! Who wants to work on this? 😉
